grr/server/grr_response_server/gui/api_plugins/metadata.py

```diff
@@ -952,3 +1003,116 @@ def _GetMapFieldKeyValueTypes(
+  root = ComponentTrieNode("", "")
+
+  for route in routes:
+    curr_node = root
+    for component in route:
+      if component not in curr_node.children:
+        curr_node.children[component] = (
+          ComponentTrieNode(component, curr_node.path)
+        )
+      curr_node = curr_node.children[component]
+
+    curr_node.is_route_end = True
+
+  return root
+
+
+def _GroupRoutesByStem(
+    curr_node: ComponentTrieNode,
+    path_args: List[ComponentTrieNode],
+    stem_node: Optional[ComponentTrieNode],
+    grouped_routes_stems: Dict[str, Dict],
+) -> None:
+  """Group routes with the same stem and extract required/optional path args."""
+  new_stem_node = stem_node
+
+  if curr_node.is_path_arg:
+    path_args.append(curr_node)
+
+  if curr_node.is_route_end:
+    if stem_node is None:
+      new_stem_node = curr_node
+    elif curr_node.is_path_arg:
+      grouped_routes_stems[stem_node.path]["optional"].append(curr_node)
+    else:
+      # It's a terminal fixed URL component => new route which might have some
+      # optional path arguments.
+      new_stem_node = curr_node
+  else:
+    # Non-terminal components that are waiting for components farther to the
+    # right to end the URL they are part of. The path arguments in this
+    # situation will be marked as required.
+    new_stem_node = None
+
+  if new_stem_node is not None and new_stem_node != stem_node:
+    # We've detected a new route end with its associated fixed stem.
+    grouped_routes_stems[new_stem_node.path] = {
+      "required": path_args.copy(),
+      "optional": [],
+    }
+
+  for child in curr_node.children.values():
+    _GroupRoutesByStem(child, path_args, new_stem_node, grouped_routes_stems)
+
+  # We'll go back to the parent of this node, so we remove the current path arg
+  # from the list of path args, as we will next go on a diferent branch/path in
+  # the trie.
+  if curr_node.is_path_arg:
+    path_args.pop()
+
+
+def _GetGroupedRoutes(routes: List[List[str]]) -> List[RouteInfo]:
+  """Get a list of routes and their required and optional path arguments.
+
+  This function creates a trie of the path components from the given list of
+  routes and then identifies groups of routes that can be reduced to a single
+  route that has optional path arguments.
+
+  Args:
+    routes: A list of routes where each route is represented as a list where
+      the first element is a string representing the HTTP method of the route
+      and the following elements are strings representing the path components.
+
+  Returns:
+    A list of `RouteInfo`s which are tuples consisting of three elements: a list
+    with the HTTP method and the components of the largest path that covers
+    multiple of the initial routes given as argument, a list of path components
+    which represent the required path arguments and a list of path components
+    which represent the optional path arguments.
```

Comment:
This shows the problem with using tuples. If you were to use a class with named fields (e.g. through `NamedTuple`), you could put these descriptions to appropriate fields making it much more obvious and better to read.

In general: avoid tuples at all costs. I don't see any value in using tuples in any production-ready code. They can be useful for quick run-and-forget scripts, though.

Reply:
You're right. I converted `RouteInfo` to be a `NamedTuple` instead of a type alias. Thank you!